algebra/group/subgroup.lean

```diff
@@ -0,0 +1,147 @@
+have h : a⁻¹ * (a * b) * a⁻¹⁻¹ ∈ S, from normal_subgroup.normal (a * b) hab a⁻¹,
+by simp at h; exact h
+
+-- Examples of subgroups
+@[simp]
+def trivial (G : Type u) [group G] : set G := {1}
+
+instance trivial_normal : normal_subgroup (trivial G) :=
+by refine {..}; simp {contextual := tt}
+
+instance univ_subgroup : subgroup (@univ G) :=
+by split; simp
+
+def center (G : Type u) [group G] : set G := {z | ∀ g, g * z = z * g}
+
+instance center_normal : normal_subgroup (center G) := {
+  one_mem := by simp [center],
+  mul_mem := begin
+  intros a b ha hb g,
+  rw [center, mem_set_of_eq] at *,
+  rw [←mul_assoc, ha g, mul_assoc, hb g, ←mul_assoc]
+  end,
+  inv_mem := begin
+  assume a ha g,
+  simp [center] at *,
+  calc
+    g * a⁻¹ = a⁻¹ * (g * a) * a⁻¹     : by simp [ha g]
+    ...     = a⁻¹ * g                 : by rw [←mul_assoc, mul_assoc]; simp
+  end,
+  normal := begin
+  simp [center, mem_set_of_eq],
+  intros n ha g h,
+  calc
+    h * (g * n * g⁻¹) = h * n               : by simp [ha g, mul_assoc]
+    ...               = g * g⁻¹ * n * h     : by rw ha h; simp
+    ...               = g * n * g⁻¹ * h     : by rw [mul_assoc g, ha g⁻¹, ←mul_assoc]
+  end
+}
+
+end subgroup
+
+-- Homomorphism subgroups
+namespace is_group_hom
+open subgroup
+variables [group G] [group H]
+
+@[simp]
+def kernel (f : G → H) [is_group_hom f] : set G := preimage f (trivial H)
+
+@[simp] lemma mem_ker_one {f : G → H} [is_group_hom f] {x : G} (h : x ∈ kernel f) : f x = 1 := by simp [kernel] at h; simp [h]
+
+@[simp] lemma one_ker_inv {f : G → H} [is_group_hom f] {a b : G} (h : f (a * b⁻¹) = 1) : f a = f b := by rw ←inv_inv (f b); rw [mul f, inv f] at h; exact eq_inv_of_mul_eq_one h
+
+@[simp] lemma inv_ker_one {f : G → H} [is_group_hom f] {a b : G} (h : f a = f b) : f (a * b⁻¹) = 1 :=
+have f a * (f b)⁻¹ = 1, by rw h; apply mul_right_inv,
+by rw [←inv f, ←mul f] at this; exact this
+
+@[simp] lemma ker_inv {f : G → H} [is_group_hom f] {a b : G} (h : a * b⁻¹ ∈ kernel f) : f a = f b := one_ker_inv $ mem_ker_one h
+
+@[simp] lemma inv_ker {f : G → H} [is_group_hom f] {a b : G} (h : f a = f b) : a * b⁻¹ ∈ kernel f := by simp [mem_set_of_eq]; exact inv_ker_one h
+
+lemma one_iff_ker_inv (f : G → H) [is_group_hom f] (a b : G) : f a = f b ↔ f (a * b⁻¹) = 1 :=
+⟨inv_ker_one, one_ker_inv⟩
+
+lemma inv_iff_ker (f : G → H) [w : is_group_hom f] (a b : G) : f a = f b ↔ a * b⁻¹ ∈ kernel f :=
+⟨@inv_ker _ _ _ _ f w _ _, @ker_inv _ _ _ _ f w _ _⟩ -- TODO: I don't understand why I can't just write ⟨inv_ker, ker_inv⟩ here. (This gives typeclass errors; it can't find `w`.)
+
+instance image_subgroup (f : G → H) [is_group_hom f] (S : set G) [subgroup S] : subgroup (f '' S) := {
+  subgroup .
+  mul_mem := assume a₁ a₂ ⟨b₁, hb₁, eq₁⟩ ⟨b₂, hb₂, eq₂⟩,
+  ⟨b₁ * b₂, mul_mem hb₁ hb₂, by simp [eq₁, eq₂, mul f]⟩,
+  one_mem := ⟨1, one_mem S, one f⟩,
+  inv_mem := assume a ⟨b, hb, eq⟩,
+  ⟨b⁻¹, inv_mem hb, by rw inv f; simp *⟩ 
+}
```

Comment:
style: braces on same line, before and after. Also leave off the `subgroup .` if it can be inferred (which should always be the case, assuming you have stated the target type of the theorem).

Reply:
Fixed.